internal/template-validator/tlsinfo/tlsinfo.go

```diff
@@ -174,31 +162,82 @@ func loadCertificates(directory string) (serverCrt *tls.Certificate, err error)
 	return &cert, nil
 }
 
-func (ti *TLSInfo) getCertificate() *tls.Certificate {
-	ti.certLock.Lock()
-	defer ti.certLock.Unlock()
-	return ti.cert
+func (ti *TLSInfo) updateTLSOptions(directory string) {
+	tlsOptions, err := loadTLSOptions(directory)
+
+	ti.lock.Lock()
+	defer ti.lock.Unlock()
+
+	if err != nil {
+		ti.minTLSVersion = 0
+		ti.cipherSuites = nil
+		ti.tlsOptionsError = err
+		logger.Log.Error(err, "failed to load TLS options",
+			"directory", directory)
+		return
+	}
+
+	if tlsOptions.IsEmpty() {
+		// Using default configuration
+		ti.minTLSVersion = 0
+		ti.cipherSuites = nil
+		ti.tlsOptionsError = nil
+		logger.Log.Info("TLS options are empty, using default configuration")
+		return
+	}
+
+	minVersion, err := tlsOptions.MinTLSVersionId()
+	if err != nil {
+		ti.minTLSVersion = 0
+		ti.cipherSuites = nil
+		ti.tlsOptionsError = fmt.Errorf("TLS Configuration broken, min version misconfigured: %w", err)
+		logger.Log.Error(ti.tlsOptionsError, "TLS options are not valid",
+			"directory", directory)
+		return
+	}
+
+	ti.minTLSVersion = minVersion
+	ti.cipherSuites = common.CipherIDs(tlsOptions.OpenSSLCipherNames, nil)
+	ti.tlsOptionsError = nil
+	logger.Log.Info("TLS options retrieved", "directory", directory)
+}
+
+func loadTLSOptions(directory string) (*common.SSPTLSOptions, error) {
+	optionsFilePath := filepath.Join(directory, TLSOptionsFilename)
+	optionsJson, err := os.ReadFile(optionsFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	tlsOptions := &common.SSPTLSOptions{}
+	if err = json.Unmarshal(optionsJson, tlsOptions); err != nil {
+		return nil, err
+	}
+
+	return tlsOptions, nil
 }
 
-func (ti *TLSInfo) CreateTlsConfig() *tls.Config {
-	tlsConfig := &tls.Config{
-		GetCertificate: func(info *tls.ClientHelloInfo) (*tls.Certificate, error) {
-			cert := ti.getCertificate()
-			if cert == nil {
-				return nil, errors.New("no server certificate, server is not yet ready to receive traffic")
-			}
-			return cert, nil
-		},
+func (ti *TLSInfo) CreateTlsConfig() (*tls.Config, error) {
+	ti.lock.Lock()
+	defer ti.lock.Unlock()
+
+	if ti.certError != nil {
+		return nil, fmt.Errorf("error getting certificate: %w", ti.certError)
 	}
 
-	if !ti.sspTLSOptions.IsEmpty() {
-		tlsConfig.CipherSuites = common.CipherIDs(ti.sspTLSOptions.OpenSSLCipherNames, nil)
-		minVersion, err := ti.sspTLSOptions.MinTLSVersionId()
-		if err != nil {
-			panic(fmt.Sprintf("TLS Configuration broken, min version misconfigured %v", err))
-		}
-		tlsConfig.MinVersion = minVersion
+	if ti.tlsOptionsError != nil {
+		return nil, fmt.Errorf("error getting TLS options: %w", ti.tlsOptionsError)
+	}
+
+	if ti.cert == nil {
+		return nil, fmt.Errorf("the TLS certificate is unexpectedly nil")
+	}
+
```

Comment:
What about `ti.cipherSuites`? Are we ever allowing them to be `nil`?

Reply:
Yes, they are set to `nil` if the config file is empty. See line 183.

Reply:
I don't understand your question. Can you explain what do you mean?

Reply:
If the config file is empty, then TLS is going to be disabled and therefore `CreateTlsConfig` is not going to get called, correct? To me that means that `CreateTlsConfig` should not accept `ti.cipherSuites` to be `nil`.

Reply:
No, TLS is always enabled. If the config file is empty, it just means that SSP did not pass any specific configuration to validator, and it should use the default.

Reply:
ACK